---
title: "uTILity: General Data Processing"
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
output: rmarkdown::html_vignette
theme: united
df_print: kable
vignette: >
  %\VignetteIndexEntry{uTILity: General Data Processing}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo=FALSE, results="hide", message=FALSE}
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
library(BiocStyle)
```

# Introduction

The original intent of assembling a data set of publicly-available tumor-infiltrating T cells (TILs) with paired TCR sequencing was to expand and improve the [scRepertoire](https://github.com/ncborcherding/scRepertoire) R package. However, after some discussion, we decided to release the data set for everyone, a complete summary of the sequencing runs and the sample information can be found in the meta data of the Seurat object.

The code below is the general purpose variant for the vast majority of samples - there are cohorts that need to be de-multiplexed or separated from combined count matrices, these can be found in **/data/processedData/combinedDataSets**.

## Citation

utility is not published nor has it gone through peer review. If utilizing the data, please cite the corresponding data sets. This can be found under **"Cohort"** in the meta data.

## Questions or Comments

Questions, comments, suggestions, please feel free to contact Nick Borcherding via this repository, [email](mailto:ncborch@gmail.com), or using [twitter](https://twitter.com/theHumanBorch). **Also** if you are interested in dimensional reduction of TCR sequencing data, please get in touch as that is the intent of the assembled data and we would love to talk more!

***

# Loading Libraries

In general I like to load libraries here that we will use universally or during the intitation process, and then call other libraries when we need them in the code chunks that are relevant. 

```{r}
suppressPackageStartupMessages(library(Seurat))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(viridis))
suppressPackageStartupMessages(library(scDblFinder))
suppressPackageStartupMessages(library(BiocParallel))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(SingleR))
suppressPackageStartupMessages(library(celldex))
suppressPackageStartupMessages(library(scRepertoire))
suppressPackageStartupMessages(library(Azimuth))
suppressPackageStartupMessages(library(SeuratData))
suppressPackageStartupMessages(library(scran))
suppressPackageStartupMessages(library(patchwork))
```

***

# Loading and Processing the Data

## Load, filter and combine Data
```{r eval=FALSE}
dir.create("./qc")

file_list <- list.files("~/Documents/GitHub/utility/data/SequencingRuns")

###########################
#Load Annotation Reference
###########################
HPCA <- HumanPrimaryCellAtlasData()
Monaco <- MonacoImmuneData()

options(Seurat.object.assay.version = "v5")

for (i in seq_along(file_list)){
    tmp <-  Read10X(paste0("~/Documents/GitHub/utility/data/SequencingRuns/", file_list[i]))
    if(inherits(tmp, "list")) {
      tmp <- tmp[[1]]
     # rownames(tmp) <- stringr::str_split(rownames(tmp), "_", simplify = TRUE)[,2]
      #tmp <- tmp[rownames(tmp) != "",]
    }
   
    
    tmp <- CreateAssayObject(counts = tmp)
    SeuratObj <-  CreateSeuratObject(counts = tmp, 
                                     assay = "RNA",
                               project = file_list[i])
    SeuratObj$nCount_RNA <- colSums(SeuratObj@assays$RNA@counts)
    SeuratObj$nFeature_RNA <- colSums(SeuratObj@assays$RNA@counts != 0)
    
    SeuratObj <-  subset(SeuratObj, subset = nFeature_RNA > 100) #filter out low count/feature cells
    SeuratObj  <- RenameCells(object = SeuratObj , new.names = paste0(file_list[i], "_", rownames(SeuratObj[[]])))
    
    SeuratObj[["mito.genes"]] <- PercentageFeatureSet(SeuratObj, pattern = "^MT-")
    SeuratObj[["ribo.genes"]] <- PercentageFeatureSet(SeuratObj, pattern = "^RPS|RPL-")
    
    VlnPlot(object = SeuratObj, 
            features = c("nCount_RNA", "nFeature_RNA", "mito.genes", "ribo.genes"), 
            pt.size = 0) + 
      theme(legend.position = "none") + 
      plot_layout(ncol =2)
    ggsave(paste0("~/Documents/GitHub/utility/outputs/qc/", file_list[i], ".pdf"), height = 8, width=8)
    
    
    ###########################
    #Here is the filtering step
    ############################
    standev <- sd(log(SeuratObj$nFeature_RNA))*2.5 #cutting off above standard deviation of 2.5
    mean <- mean(log(SeuratObj$nFeature_RNA))
    cut <- round(exp(standev+mean))
    SeuratObj <- subset(SeuratObj, subset = mito.genes < 10 & nFeature_RNA < cut)
    
    ###########################################
    #Estimate Doublets for Each Sequencing Run
    ############################################
    sce <- as.SingleCellExperiment(SeuratObj)
    sce <- scDblFinder(sce)
    doublets <- data.frame(db.class = sce$scDblFinder.class, db.score = sce$scDblFinder.score)
    rownames(doublets) <- rownames(sce@colData)
    SeuratObj <- AddMetaData(SeuratObj, doublets)

    
    ###########################################
    #Seurat Azimuth Annotation
    ############################################
    SeuratObj <- NormalizeData(SeuratObj, verbose = FALSE)
    SeuratObj <- ScaleData(SeuratObj, verbose = FALSE)
    VariableFeatures(SeuratObj) <- getTopHVGs(as.SingleCellExperiment(SeuratObj), 
                                        n=2000)
    SeuratObj <- RunPCA(SeuratObj, verbose = FALSE)
    SeuratObj<- RunUMAP(SeuratObj, dims = 1:30, verbose = FALSE)
    SeuratObj <- FindNeighbors(object = SeuratObj, 
                         features = VariableFeatures(SeuratObj), 
                         verbose = FALSE)
   
   SeuratObj <- RunAzimuth(SeuratObj, 
                      reference = "pbmcref",
                      verbose = FALSE)
      
    #############################################
    #Singler Annotation of Cell Types
    #############################################
      
    com.res1 <- SingleR(sce, ref=HPCA, labels=HPCA$label.fine, assay.type.test=1)
  
    df <- data.frame("labels" = com.res1$labels, "pruned.labels" = com.res1$pruned.labels)
    rownames(df) <- rownames(com.res1)
    colnames(df) <- paste0("HPCA.", colnames(df))
    SeuratObj <- AddMetaData(SeuratObj,  df)
  
    
    com.res2 <- SingleR(sce, ref=Monaco, labels=Monaco$label.fine, assay.type.test=1)
    df <- data.frame("labels" = com.res2$labels, "pruned.labels" = com.res2$pruned.labels)
    rownames(df) <- rownames(com.res1)
    colnames(df) <- paste0("Monaco.", colnames(df))
    SeuratObj <- AddMetaData(SeuratObj,  df)
    rm(df)
    rm(sce)
      
    ######################################
    #Adding TCR clonotypes
    ######################################
    
    TCR.file <- list.files(paste0("~/Documents/GitHub/utility/data/SequencingRuns/", file_list[i]), pattern = "annotations")[1]
    if(is.na(TCR.file)) {
      SeuratObj$CTaa <- NA
      SeuratObj$CTnt <- NA
      SeuratObj$CTgene <- NA
      SeuratObj$CTstrict <- NA
      SeuratObj$clonalProportion <- NA
      SeuratObj$clonalFrequency <- NA
      SeuratObj$cloneSize <- NA
    } else {
      TCR.file <- read.csv(paste0("~/Documents/GitHub/utility/data/SequencingRuns/", file_list[i], "/", TCR.file))
      combinedObject <- combineTCR(TCR.file, 
                                   samples = file_list[i], 
                                   filterMulti = TRUE)
      SeuratObj <- combineExpression(combinedObject, 
                               SeuratObj, 
                               cloneCall = "aa")
    }
    
    SeuratObj <- DietSeurat(SeuratObj)
    SeuratObj[["prediction.score.celltype.l1"]] <- NULL
    SeuratObj[["prediction.score.celltype.l2"]] <- NULL
    SeuratObj[["prediction.score.celltype.l3"]] <- NULL
    SeuratObj[["query_ref.nn"]] <- NULL
    SeuratObj@assays$RNA$scale.data <- NULL
    
    #################################
    #Saving Preliminary Seurat Object
    #################################
    saveRDS(SeuratObj, paste0("~/Documents/GitHub/utility/data/processedData/seuratObjects/", file_list[i], ".rds"))
    rm(SeuratObj)
    rm(tmp)
    gc()
}
```

# Conclusions

The preceding was all of the code generated to produce the final seurat object. To reiterate - questions, comments, suggestions, please feel free to contact Nick Borcherding via this repository, [email](mailto:ncborch@gmail.com), or using [twitter](https://twitter.com/theHumanBorch). **Also** if you are interested in dimensional reduction of TCR sequencing data, please get in touch as that is the intent of the assembled data and we would love to talk more!

```{r}
writeLines(capture.output(sessionInfo()), "/summaryInfo/sessionInfo.txt")
```


