---
title: Processing Combined Data: GSE179994
author: 
- name: Nick Borcherding
  email: ncborch@gmail.com
  affiliation: Washington University in St. Louis, School of Medicine, St. Louis, MO, USA
date: "March 3rd, 2022"
output:
  BiocStyle::html_document:
    toc_float: true

---

```{r, echo=FALSE, results="hide", message=FALSE}
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
library(BiocStyle)
```

#Utility Workflow and Export

```{r}
suppressPackageStartupMessages(library(Seurat))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(viridis))
suppressPackageStartupMessages(library(scDblFinder))
suppressPackageStartupMessages(library(BiocParallel))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(SingleR))
suppressPackageStartupMessages(library(celldex))
suppressPackageStartupMessages(library(scRepertoire))
suppressPackageStartupMessages(library(Azimuth))
suppressPackageStartupMessages(library(SeuratData))
suppressPackageStartupMessages(library(scran))
suppressPackageStartupMessages(library(patchwork))
```


#RNA processing 

```{r}
RNA.dat <- Matrix::readMM(gzfile("GSE181061_ccRCC_4pt_scRNAseq_CD45plus_matrix.mtx.gz"))
genes <- read.table(gzfile("GSE181061_ccRCC_4pt_scRNAseq_CD45plus_genes.tsv.gz"))
barcodes <- read.table(gzfile("GSE181061_ccRCC_4pt_scRNAseq_CD45plus_barcodes.tsv.gz"))
RNA.dat@Dimnames[[1]] <- genes$V1
RNA.dat@Dimnames[[2]] <- barcodes$V1

options(Seurat.object.assay.version = "v5")
RNA.dat <-  CreateSeuratObject(RNA.dat)
RNA.dat$nCount_RNA <- colSums(RNA.dat@assays$RNA@layers$counts)
RNA.dat$nFeature_RNA <- colSums(RNA.dat@assays$RNA@layers$counts != 0)

sample <- stringr::str_split(rownames(RNA.dat[[]]), "[+]", simplify = TRUE)[,1]

RNA.dat$orig.ident <- sample

list.files <- list.files("./contigs")

contig.list <- lapply(list.files, FUN = function(x) {
  read.csv(paste0("./contigs/", x))
  })

combinedObject <- combineTCR(contig.list, 
                             samples = paste0(unique(RNA.dat$orig.ident), "+"), 
                             filterMulti = TRUE)

for (i in seq_along(combinedObject)) {
  combinedObject[[i]]$barcode <- stringr::str_remove_all(combinedObject[[i]]$barcode, "-1")
}


RNA.dat <- combineExpression(combinedObject, RNA.dat, cloneCall = "aa")
```

# Getting Porcessed Sequencing Runs

```{r include=FALSE}
HPCA <- HumanPrimaryCellAtlasData()
Monaco <- MonacoImmuneData()

directory <- readxl::read_xlsx("~/Documents/GitHub/utility/summaryInfo/sample.directory.xlsx") #Meta.data
directory <- directory[directory$Cohort == "GSE181061",]

utility.sample.id <- directory$SampleLabel


list <- SplitObject(RNA.dat, split.by = "orig.ident")

for (i in seq_along(list)){
    tmp <- list[[i]]

    tmp <- subset(tmp, subset = nFeature_RNA > 100) #filter out low count/feature cells\
    new.barcodes <- rownames(tmp@meta.data)
    new.barcodes <- stringr::str_sub(new.barcodes, nchar(new.barcodes)-15, nchar(new.barcodes))
    if(length(which(duplicated(new.barcodes))) > 0) {
      pos.duplicated <- which(duplicated(new.barcodes))
      #Subset Seurat Object
      keep.cells <- rownames(tmp@meta.data)[-pos.duplicated]
      tmp <- subset(tmp, cells = keep.cells)
      #Remove Stripped Barcodes
      new.barcodes <- new.barcodes[-pos.duplicated]
    }
    
    ut.sample <- utility.sample.id[i]
    
    tmp  <- RenameCells(object = tmp , new.names = paste0(ut.sample, "_", new.barcodes))
    tmp$orig.ident <- ut.sample

    tmp[["mito.genes"]] <- PercentageFeatureSet(tmp, pattern = "^MT-")
    tmp[["ribo.genes"]] <- PercentageFeatureSet(tmp, pattern = "^RPS|RPL-")
    
    VlnPlot(object = tmp, 
            features = c("nCount_RNA", "nFeature_RNA", "mito.genes", "ribo.genes"), 
            pt.size = 0) + 
      theme(legend.position = "none") + 
      plot_layout(ncol =2)
    ggsave(paste0("~/Documents/GitHub/utility/qc/", ut.sample, ".pdf"), height = 8, width=8)
    
    ###########################
    #Here is the filtering step
    ############################
    standev <- sd(log(tmp$nFeature_RNA))*2.5 #cutting off above standard deviation of 2.5
    mean <- mean(log(tmp$nFeature_RNA))
    cut <- round(exp(standev+mean))
    tmp <- subset(tmp, subset = mito.genes < 10 & nFeature_RNA < cut)
    
    ###########################################
    #Estimate Doublets for Each Sequencing Run
    ############################################
    sce <- as.SingleCellExperiment(tmp)
    sce <- scDblFinder(sce, BPPARAM=MulticoreParam(3))
    doublets <- data.frame(db.class = sce$scDblFinder.class, db.score = sce$scDblFinder.score)
    rownames(doublets) <- rownames(sce@colData)
    tmp <- AddMetaData(tmp, doublets)

    
    ###########################################
    #Seurat Azimuth Annotation
    ############################################
    tmp <- NormalizeData(tmp, verbose = FALSE)
    tmp <- ScaleData(tmp, verbose = FALSE)
    VariableFeatures(tmp) <- getTopHVGs(as.SingleCellExperiment(tmp), 
                                        n=2000)
    tmp <- RunPCA(tmp, verbose = FALSE)
    tmp<- RunUMAP(tmp, dims = 1:30, verbose = FALSE)
    tmp <- FindNeighbors(object = tmp, 
                         features = VariableFeatures(tmp), 
                         verbose = FALSE)
   
   tmp <- RunAzimuth(tmp, 
                      reference = "pbmcref",
                      verbose = FALSE)
      
    #############################################
    #Singler Annotation of Cell Types
    #############################################
      
    com.res1 <- SingleR(sce, ref=HPCA, labels=HPCA$label.fine, assay.type.test=1)
    df <- data.frame("labels" = com.res1$labels, "pruned.labels" = com.res1$pruned.labels)
    rownames(df) <- rownames(com.res1)
    colnames(df) <- paste0("HPCA.", colnames(df))
    tmp <- AddMetaData(tmp,  df)
  
    
    com.res2 <- SingleR(sce, ref=Monaco, labels=Monaco$label.fine, assay.type.test=1)
    df <- data.frame("labels" = com.res2$labels, "pruned.labels" = com.res2$pruned.labels)
    rownames(df) <- rownames(com.res1)
    colnames(df) <- paste0("Monaco.", colnames(df))
    tmp <- AddMetaData(tmp,  df)
    rm(df)
    rm(sce)
    
    tmp <- DietSeurat(tmp)
    tmp[["prediction.score.celltype.l1"]] <- NULL
    tmp[["prediction.score.celltype.l2"]] <- NULL
    tmp[["prediction.score.celltype.l3"]] <- NULL
    tmp[["query_ref.nn"]] <- NULL
    tmp@assays$RNA$scale.data <- NULL
    
    saveRDS(tmp, paste0("~/Documents/GitHub/utility/data/processedData/seuratObjects/", ut.sample, ".rds"))
    rm(tmp)
    gc()
}
```